### PR TITLE
Issue 15800 file assets identifier enhancements

### PR DESCRIPTION
**Pull request by fabrizzio-dotCMS**

### CODE REVIEW

**Review by jdotcms — COMMENTED** [review]
Very nice

File: dotCMS/src/main/java/com/dotmarketing/portlets/fileassets/business/FileAssetAPIImpl.java [hunk 1]
```diff
@@ -304,35 +292,45 @@ public boolean fileNameExists(Host host, Folder folder, String fileName, String
...
+		   return false;
+		}
+            //Beyond this point we know something matches that path for that host.
+			if(!UtilMethods.isSet(identifier)){
+			   //it's a brand new contentlet we're dealing with
+			   //At this point we know it DOES exist, and since we're dealing with a fresh contentlet that hasn't even been inserted yet (We don't need to worry about lang).
+			   return true;
+			} else {
+				// old logic. ie calling fileNameExists method without languageId parameter.
+				if (languageId == -1){
```
Comment (jgambarios) [inline dotCMS/src/main/java/com/dotmarketing/portlets/fileassets/business/FileAssetAPIImpl.java]: Why do we care about the language, `languageId` should not even be a parameter, we just care about what is in the identifier table right?

File: dotCMS/src/main/java/com/dotmarketing/portlets/fileassets/business/FileAssetAPIImpl.java [hunk 2]
```diff
@@ -304,35 +292,45 @@ public boolean fileNameExists(Host host, Folder folder, String fileName, String
...
+				}
+                // Now we have an identifier and a lang.
+				// if the file-asset identifier is different from the contentlet identifier we're looking at. Then it does exist already.
+				if (!identifier.equals(fileAssetIdentifier.getId())) {
+				   // the path's already associated with another contentlet. Get out of here.
+				   return true;
+				} else {
+				    // if we're looking at the same conentlet we need to verify if the file-Name's already associated with the language we got.
+					try {
+						return (null != contAPI.findContentletByIdentifier(fileAssetIdentifier.getId(), false, languageId, APILocator.getUserAPI().getSystemUser(), false));
```
Comment (jgambarios) [inline dotCMS/src/main/java/com/dotmarketing/portlets/fileassets/business/FileAssetAPIImpl.java]: Why we need to look for the contentlet in order if a assent name was already taken?

**Review by jgambarios — CHANGES_REQUESTED** [review]

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java [hunk 3]
```diff
@@ -299,11 +299,27 @@ public Contentlet findContentletForLanguage(long languageId, Identifier contentl
...
+    @CloseDBIfOpened
+    @Override
+    public List<Contentlet> findContentletsForAllLanguages(final Identifier contentletId) {
+        final LanguageAPI languageAPI = APILocator.getLanguageAPI();
+        final List<Language> allLanguages = languageAPI.getLanguages();
+        return allLanguages.stream().map(language -> {
+            try {
+                return findContentletForLanguage(language.getId(), contentletId);
+            } catch (DotDataException | DotSecurityException e) {
+                Logger.debug(this, "No working contentlet found for language: "+ language.toString());
```
[the remark this answers is not in the document]
Comment (jcastro-dotcms) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java]: Yes please, very important for Support tasks.

File: dotCMS/src/main/java/com/dotmarketing/portlets/fileassets/business/FileAssetAPIImpl.java [hunk 4]
```diff
@@ -295,7 +295,7 @@ public boolean fileNameExists(final Host host, final Folder folder, final String
 		final Identifier folderId = APILocator.getIdentifierAPI().find(folder);
 		final String path = folder.getInode().equals(FolderAPI.SYSTEM_FOLDER) ? StringPool.FORWARD_SLASH + fileName : folderId.getPath() + fileName;
 		final Identifier fileAssetIdentifier = APILocator.getIdentifierAPI().find(host, path);
-		if(( null == fileAssetIdentifier || !InodeUtils.isSet(fileAssetIdentifier.getId()) || "folder".equals(fileAssetIdentifier.getAssetType()))){
+		if( null == fileAssetIdentifier || !InodeUtils.isSet(fileAssetIdentifier.getId()) || "folder".equals(fileAssetIdentifier.getAssetType())){
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/fileassets/business/FileAssetAPIImpl.java]: I think you can use something like `Folder.getType()` to not use the String "folder" directly?

**Review by jcastro-dotcms — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java [hunk 5]
```diff
@@ -299,11 +299,27 @@ public Contentlet findContentletForLanguage(long languageId, Identifier contentl
...
+            try {
+                return findContentletForLanguage(language.getId(), identifier);
+            } catch (DotDataException | DotSecurityException e) {
+                Logger.debug(this, ()->String.format("No working contentlet found for language: %s and identifier: %s ", language.toString(), null != identifier ? identifier.getId() : "Unkown"));
+                return null;
+            }
+        }).filter(Objects::nonNull).collect(CollectionsUtils.toImmutableList());
+    }
+
+
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java]: I am pretty sure we have this method in a few places and don't need to add to the already enormous contentletAPI - see:
`findAllVersions`
`getAllLanguages`

**Review by wezell — COMMENTED** [review]
See my feedback

**Review by wezell — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/FileAssetsResource.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+import com.dotmarketing.portlets.contentlet.model.Contentlet;
+import com.dotmarketing.portlets.contentlet.model.ResourceLink;
+import com.dotmarketing.portlets.contentlet.model.ResourceLink.ResourceLinkBuilder;
+import com.dotmarketing.util.Logger;
+import com.dotmarketing.util.UtilMethods;
+import com.google.common.collect.ImmutableMap;
+import com.liferay.portal.model.User;
+import javax.servlet.http.HttpServletRequest;
+
+@Path("/v1/content/asset")
```
Comment (jgambarios) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/FileAssetsResource.java]: Please change the **path** and **package** to use also `FileAsset`

**Review by jgambarios — CHANGES_REQUESTED** [review]

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/FileAssetsResource.java (hunk 6 above)
Comment (jgambarios) [reply dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/FileAssetsResource.java]: Also, taking with Daniel should be in plural, something like `@Path("/v1/content/fileassets")`

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/FileAssetsResource.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+     * @throws DotStateException
+     * @throws DotSecurityException
+     */
+    @GET
+    @JSONP
+    @NoCache
+    @Path("/resourcelink")
+    @Produces(MediaType.APPLICATION_JSON + ";charset=UTF-8")
+    public Response findResourceLink(@Context final HttpServletRequest request,
+            @QueryParam("inode") final String inode) throws DotStateException {
```
Comment (jgambarios) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/FileAssetsResource.java]: The inode should be a path param instead a query param, something like `@Path("/{inode}/resourcelink")`

**Review by jgambarios — CHANGES_REQUESTED** [review]

**Review by jdotcms — APPROVED** [review]

**Review by jgambarios — APPROVED** [review]

**Pull request merged**